File: src/main/java/com/app/wydatki/service/BudgetService.java
package com.app.wydatki.service;

import com.app.wydatki.dto.BudgetDTO;
import com.app.wydatki.model.Budget;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface BudgetService {
    
    Budget createBudget(BudgetDTO budgetDTO, String userEmail);
    
    Budget updateBudget(Long id, BudgetDTO budgetDTO, String userEmail);
    
    void deleteBudget(Long id, String userEmail);
    
    Optional<Budget> getBudgetById(Long id, String userEmail);
    
    List<Budget> getUserBudgets(String userEmail);
    
    List<Budget> getActiveBudgets(String userEmail);
    
    List<Budget> getUpcomingBudgets(String userEmail);
    
    List<Budget> getPastBudgets(String userEmail);
    
    List<Budget> getBudgetsInDateRange(String userEmail, LocalDate startDate, LocalDate endDate);
    
    BigDecimal getTotalBudgetAmount(String userEmail, LocalDate startDate, LocalDate endDate);
    
    BigDecimal getRemainingBudgetAmount(Long budgetId, String userEmail);
    
    Budget addSpending(Long budgetId, BigDecimal amount, String userEmail);
    
    Budget removeSpending(Long budgetId, BigDecimal amount, String userEmail);
    
    boolean isBudgetActive(Long budgetId, String userEmail);
    
    List<Budget> getBudgetsByCategory(String userEmail, String category);
}

File: frontend/src/services/budget.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../environments/environment';

export interface Budget {
  id: number;
  name: string;
  amount: number;
  spentAmount: number;
  startDate: string;
  endDate: string;
  description: string;
  category: string;
}

@Injectable({
  providedIn: 'root'
})
export class BudgetService {
  private apiUrl = `${environment.apiUrl}/budgets`;

  constructor(private http: HttpClient) {}

  getAllBudgets(): Observable<Budget[]> {
    return this.http.get<Budget[]>(this.apiUrl);
  }

  getActiveBudgets(): Observable<Budget[]> {
    return this.http.get<Budget[]>(`${this.apiUrl}/active`);
  }

  getUpcomingBudgets(): Observable<Budget[]> {
    return this.http.get<Budget[]>(`${this.apiUrl}/upcoming`);
  }

  getPastBudgets(): Observable<Budget[]> {
    return this.http.get<Budget[]>(`${this.apiUrl}/past`);
  }

  getBudgetsByCategory(category: string): Observable<Budget[]> {
    return this.http.get<Budget[]>(`${this.apiUrl}/category/${category}`);
  }

  createBudget(budget: Partial<Budget>): Observable<Budget> {
    return this.http.post<Budget>(this.apiUrl, budget);
  }

  updateBudget(id: number, budget: Partial<Budget>): Observable<Budget> {
    return this.http.put<Budget>(`${this.apiUrl}/${id}`, budget);
  }

  deleteBudget(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  addSpending(id: number, amount: number): Observable<Budget> {
    return this.http.post<Budget>(`${this.apiUrl}/${id}/add-spending`, null, {
      params: { amount: amount.toString() }
    });
  }

  removeSpending(id: number, amount: number): Observable<Budget> {
    return this.http.post<Budget>(`${this.apiUrl}/${id}/remove-spending`, null, {
      params: { amount: amount.toString() }
    });
  }
}
